### PR TITLE
StrongRoute QSP

### PR DESCRIPTION
**Pull request by Allcharles**
# Strong Route Query Parameters

Added tracking of query parameters to the `StrongRoute`. This will allow for links in the action/secondary menus to handle dynamic parameters.

## Changes

- Added query parameters to `StrongRoute`
- Updated `MenuComponent` and `InternalLinkComponent` to handle applying `StrongRoute` query params to the link

## Final Checklist

- [x] Assign reviewers if you have permission
- [x] Assign labels if you have permission
- [x] Link issues related to PR
- [x] Ensure project linter is not producing any warnings (`npm run lint`)
- [x] Ensure build is passing on all browsers (`npm run test:all`)
- [ ] Ensure CI build is passing
- [ ] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

### CODE REVIEW

File: src/app/components/security/pages/login/login.component.ts [hunk 1]
```diff
@@ -106,6 +107,8 @@ class LoginComponent extends FormTemplate<LoginDetails> implements OnInit {
   protected redirectUser() {
     if (this.redirectBack) {
       this.location.back();
+    } else if (this.redirectUrl instanceof StrongRoute) {
+      this.router.navigateByUrl(this.redirectUrl.toString());
```
Comment (Allcharles) [inline src/app/components/security/pages/login/login.component.ts]: Add a test for this scenario if none specifically exists

File: src/app/components/shared/menu/internal-link/internal-link.component.spec.ts [hunk 2]
```diff
@@ -47,6 +48,7 @@ describe("MenuInternalLinkComponent", () => {
         [placement]="placement"
         [tooltip]="tooltip"
         [route]="route"
+        [qsp]="qsp"
```
Comment (Allcharles) [inline src/app/components/shared/menu/internal-link/internal-link.component.spec.ts:46]: Maybe a better name would be `[params]`? Maybe investigate to check that the RouterLink directive doesn't interfere with using `[queryParams]`

File: src/app/interfaces/strongRoute.ts [hunk 3]
```diff
@@ -1,47 +1,64 @@
...
    * @param config Additional router configurations
    * @param isRoot Is this a root StrongRoute
    */
   private constructor(
-    parent: StrongRoute,
-    name: string,
-    config: Partial<Route>,
+    parent: StrongRoute = undefined,
+    name: string = StrongRoute.rootPath,
+    qsp: QSP = () => ({}),
```
Comment (Allcharles) [inline src/app/interfaces/strongRoute.ts]: Missing description

File: src/app/interfaces/strongRoute.ts [hunk 4]
```diff
@@ -55,49 +72,47 @@ export class StrongRoute {
...
+    return new StrongRoute(this, name, qsp, config, true);
   }
 
   /**
    * Method used for templating a route with parameters.
    * Use this in a template like so:
-   * <a [routerlink]="route.Format({projectId: 1, siteId: 1})" />
+   * <a [routerlink]="route.format({projectId: 1, siteId: 1})" />
    */
   public format(args: { [key: string]: string | number }): string {
```
Comment (Allcharles) [inline src/app/interfaces/strongRoute.ts]: I would like to see this function merged into `toRoute()` and `toString()` as both are effectively useless if the StrongRoute has parameters

File: src/app/components/shared/menu/internal-link/internal-link.component.spec.ts (hunk 2 above)
Comment (atruskie) [reply src/app/components/shared/menu/internal-link/internal-link.component.spec.ts:46]: this is one of the rare occasions I don't mind the abbreviation so much

also there should be no interfereance with router link

File: src/app/components/shared/menu/internal-link/internal-link.component.ts [hunk 5]
```diff
@@ -34,6 +36,7 @@ export class MenuInternalLinkComponent implements OnChanges {
   @Input() public id: string;
   @Input() public link: MenuRoute;
   @Input() public route: string;
+  @Input() public qsp: Params;
```
Comment (atruskie) [inline src/app/components/shared/menu/internal-link/internal-link.component.ts]: re last comment: if any name change, i'd change it so that it was the same as routerLink's, i.e. `queryParams`, if they behave the same way

File: src/app/helpers/advancedTypes.ts [hunk 6]
```diff
@@ -31,3 +31,6 @@ export type MayBeAsync<T> = T | Promise<T>;
 export type KeysOfType<T, TProp> = {
   [P in keyof T]: T[P] extends TProp ? P : never;
 }[keyof T];
+
+/** Sets type to either be T or null */
+export type Potential<T> = T | null;
```
Comment (atruskie) [inline src/app/helpers/advancedTypes.ts]: rename this to `Option` - it's a common functional concept and we'll do better in the long run if we reuse name

https://en.wikipedia.org/wiki/Option_type

https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/options

File: src/app/interfaces/strongRoute.ts [hunk 7]
```diff
@@ -1,47 +1,64 @@
...
-  public readonly parent: StrongRoute;
-  public readonly name: string;
-  public readonly isParameter: boolean;
-  public readonly fullRoute: string;
-  public readonly children: StrongRoute[] = [];
+  private static readonly rootPath = "";
+
+  /** Page component associated with this strong route */
+  public pageComponent: Potential<Type<PageComponent>>;
+  /** Root strong route of this strong route */
```
Comment (atruskie) [inline src/app/interfaces/strongRoute.ts]: Good docs do more than restate what is in the name of the documented element😸 

Maybe add why this is useful? E.g. allows us to form absolute routes/paths even if we only have access to current node.

Ditto this comment to all documented elements in this file

File: src/app/interfaces/strongRoute.ts [hunk 8]
```diff
@@ -1,47 +1,64 @@
...
+  public readonly root: StrongRoute;
+  /** Parent strong route of this strong route */
+  public readonly parent?: StrongRoute;
+  /** Route path/name to add/create in the strong route tree */
+  public readonly name?: string;
+  /** Route query parameters */
+  private readonly qsp?: QSP;
+  /** Is this strong route a parameter (ie. :siteId) */
```
Comment (atruskie) [inline src/app/interfaces/strongRoute.ts]:
```suggestion
  /** Is this strong route segment a route parameter? (ie. :siteId) */
```

File: src/app/interfaces/strongRoute.ts [hunk 9]
```diff
@@ -1,47 +1,64 @@
...
   private readonly full: StrongRoute[];
+  /** Additional configuration options to append to this strong route */
   private readonly config: Partial<Route>;
 
   /**
    * Constructor
    *
    * @param parent Components parent
-   * @param name Route name
+   * @param name Route name (do not include '/')
```
Comment (atruskie) [inline src/app/interfaces/strongRoute.ts]: should make this an actual validation as well as a doc

File: src/app/interfaces/strongRoute.ts [hunk 10]
```diff
@@ -1,47 +1,64 @@
...
+  /** Is this strong route a parameter (ie. :siteId) */
+  private readonly isParameter: boolean;
+  /** Children strong routes of this strong route */
+  private readonly children: StrongRoute[] = [];
+  /** List of parameter strong routes in this strong routes path */
   private readonly parameters: StrongRoute[];
+  /** List of parent strong routes (including current strong route) */
   private readonly full: StrongRoute[];
+  /** Additional configuration options to append to this strong route */
   private readonly config: Partial<Route>;
```
Comment (atruskie) [inline src/app/interfaces/strongRoute.ts]: might do with a better name. Something like `angularRouteConfig` ?

File: src/app/interfaces/strongRoute.ts [hunk 11]
```diff
@@ -1,47 +1,64 @@
...
-    parent: StrongRoute,
-    name: string,
-    config: Partial<Route>,
+    parent: StrongRoute = undefined,
+    name: string = StrongRoute.rootPath,
+    qsp: QSP = () => ({}),
+    config: Partial<Route> = {},
     isRoot?: boolean
   ) {
     this.root = this;
```
Comment (atruskie) [inline src/app/interfaces/strongRoute.ts]: it'd be good to not set this here, but instead set it down in the else case of the root check below

File: src/app/interfaces/strongRoute.ts [hunk 12]
```diff
@@ -55,49 +72,47 @@ export class StrongRoute {
...
+      ...config,
+    };
   }
 
   /**
    * A base level (root) route component
    */
   public static get base() {
-    return new StrongRoute(null, null, {});
+    return new StrongRoute();
```
Comment (atruskie) [inline src/app/interfaces/strongRoute.ts]: is it intended to more than one base? If so this is a constructor helper and it should be a method

If not, this should not return a strong route every time it is called as a property

File: src/app/interfaces/strongRoute.ts [hunk 13]
```diff
@@ -136,17 +151,24 @@ export class StrongRoute {
   /**
    * Compile the list of routes for a module
    *
-   * @param callback Callback function (usually: GetRouteConfigForPage)
+   * @param callback Callback function (usually: getRouteConfigForPage)
```
Comment (atruskie) [inline src/app/interfaces/strongRoute.ts]: Add: allows our 'pages' to add extra route data or modifications when they're set up

File: src/app/interfaces/strongRoute.ts [hunk 14]
```diff
@@ -136,17 +151,24 @@ export class StrongRoute {
...
 
     const sortRoutes = (a: Route, b: Route): -1 | 0 | 1 => {
-      if (a.path === null && b.path === null) {
+      // Root route wins
+      if (a.path === StrongRoute.rootPath || b.path === StrongRoute.rootPath) {
+        return a.path === StrongRoute.rootPath ? -1 : 1;
+      }
+
+      const aPathExists = isInstantiated(a.path);
+      const bPathExists = isInstantiated(b.path);
```
Comment (atruskie) [inline src/app/interfaces/strongRoute.ts]: `Exists` seems misleading the paths/routes may exist

Maybe `IsInstantiated`

File: src/app/interfaces/strongRoute.ts [hunk 15]
```diff
@@ -192,18 +221,37 @@ export class StrongRoute {
    * Example output: "/home/house"
    */
   public toString(): string {
-    return "/" + (this.fullRoute ? this.fullRoute : "");
+    // Prevent double '/' at start of route string
+    const path = this.toRoute().join("/");
+    return path.length >= 2 ? path.substr(1) : path;
```
Comment (atruskie) [inline src/app/interfaces/strongRoute.ts]: maybe better expressed with 
```suggestion
    return path.startsWith('/')  ? path.substr(1) : path;
```

File: src/app/interfaces/strongRoute.ts [hunk 16]
```diff
@@ -192,18 +221,37 @@ export class StrongRoute {
    * Example output: "/home/house"
    */
   public toString(): string {
```
Comment (atruskie) [inline src/app/interfaces/strongRoute.ts]: Add doc.

toString returns a single string of all route components joined together. Useful for wherewe want to see a simple diagnostic representation of the strong route.

Should not be used in router links

example output: `/projects/:projectId/sites/:siteid?filter_name=:name`

File: src/app/interfaces/strongRoute.ts (hunk 4 above)
Comment (atruskie) [reply src/app/interfaces/strongRoute.ts]: Yeah, no.

Format returns a whole, templated route (i.e. no missing parameters). It serves a distinct purpose, different from the other two methods

File: src/app/interfaces/strongRoute.ts [hunk 17]
```diff
@@ -192,18 +221,37 @@ export class StrongRoute {
...
 
   /**
-   * Router representation of the route
-   * ! This will use a path relative to the current page if directly
-   * inserted into the [routerLink] directive
+   * Router representation of the route inserted into the
+   * [routerLink] directive
    *
-   * Example output: ["home", "house"]
+   * Example output: ["/", "home", "house"]
```
Comment (atruskie) [inline src/app/interfaces/strongRoute.ts]: Add:

Outputs in the routerlink @Input format, i.e. an array of args passed to `Router#createUrlTree` via `routerLink`. See more https://angular.io/api/router/RouterLink

example output: `['/', 'projects', ':projectId', 'sites', ':siteid']`

to use queryPararms call the `queryParams` method

File: src/app/interfaces/strongRoute.ts [hunk 18]
```diff
@@ -192,18 +221,37 @@ export class StrongRoute {
...
-    return this.full.slice(1).map((x) => x.name);
+    if (this.full.length > 0) {
+      const fullRoute = this.full
+        .map((x) => x.name)
+        .filter((fragment) => !!fragment);
+
+      return ["/", ...fullRoute];
+    }
+
+    return ["/"];
```
Comment (atruskie) [inline src/app/interfaces/strongRoute.ts]: don't we have a string constant for `"/"`?

File: src/app/interfaces/strongRoute.ts [hunk 19]
```diff
@@ -192,18 +221,37 @@ export class StrongRoute {
...
    *
-   * Example output: ["home", "house"]
+   * Example output: ["/", "home", "house"]
    */
   public toRoute(): string[] {
-    return this.full.slice(1).map((x) => x.name);
+    if (this.full.length > 0) {
+      const fullRoute = this.full
+        .map((x) => x.name)
+        .filter((fragment) => !!fragment);
```
Comment (atruskie) [inline src/app/interfaces/strongRoute.ts]: what's being filtered out here?

Do we have many names that are null?

File: src/app/interfaces/strongRoute.ts (hunk 16 above)
Comment (atruskie) [reply src/app/interfaces/strongRoute.ts]: Make sure QSP are represented

File: src/app/interfaces/strongRoute.ts (hunk 4 above)
Comment (atruskie) [reply src/app/interfaces/strongRoute.ts]: Example output:

`/projects/123/sites/456?filter_name=abc`

Comment (atruskie) [reply src/app/interfaces/strongRoute.ts]: Add argument to accept QSPs?

File: src/app/interfaces/strongRoute.ts (hunk 17 above)
Comment (atruskie) [reply src/app/interfaces/strongRoute.ts]: Note there's nothing we can do here to validate if query params are being used or not.

When a user uses this in a routerLink, the routerLink args are passed to `@Input` `routerLinks` and the query params are passed separately to `queryParams`. These are two separate options and we can't control/validate that they both occur in the template

Comment (atruskie) [reply src/app/interfaces/strongRoute.ts]: What we could do is make our own `strongRouterLink` which ensures QSPs are provided or else fails.

We could also advise to always use `format` to ensure QSPs are formatted

File: src/app/interfaces/strongRoute.ts [hunk 20]
```diff
@@ -192,18 +221,37 @@ export class StrongRoute {
...
+    return ["/"];
+  }
+
+  /**
+   * Query parameters which should be supplied with this route
+   * and should be inserted into the [queryParams] directive
+   *
+   * @param params Route data parameters
+   */
+  public queryParams(params: Params) {
```
Comment (atruskie) [inline src/app/interfaces/strongRoute.ts]: add usage example:

`<a [routerlink]="siteRoute.toRoute()" queryParams="siteRoute.qsp({filer_name: abc})" />`

File: src/app/interfaces/strongRoute.ts [hunk 21]
```diff
@@ -192,18 +221,37 @@ export class StrongRoute {
...
+    return this.qsp(params);
```
Comment (atruskie) [inline src/app/interfaces/strongRoute.ts]: Does this do any validation?

**Review by atruskie — CHANGES_REQUESTED** [review]
This got complex in a few ways.

Near the end I realised I made a few erroneous assumptions as to the purpose of the some of the strong route functions.

We should meet and clarify

**Comment by Allcharles** [comment]
Image of comments:
image

File: src/app/directives/strongRoute/strong-route.directive.ts [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+  /**
+   * Additional query parameters to apply to the StrongRoute. By
+   * default, all of the angular route parameters are already given
+   * to the StrongRoute.
+   */
+  @Input() public queryParams: Params;
+
+  /** Points to original classes 'router' */
+  private _router: Router;
+  /** Points to original classes 'route' */
```
Comment (atruskie) [inline src/app/directives/strongRoute/strong-route.directive.ts]: these comments don't add much value

File: src/app/directives/strongRoute/strong-route.directive.spec.ts [hunk 23]
```diff
@@ -0,0 +1,228 @@
...
+  });
+
+  // TODO Current implementation does not work with routerLinkActive
+  xit("should not interfere with routerLinkActive", () => {});
+
+  describe("strongRoute", () => {
+    it("should handle undefined strongRoute", () => {
+      setup(undefined);
+      spec.detectChanges();
+      expect(spec.directive).toBeTruthy();
```
Comment (atruskie) [inline src/app/directives/strongRoute/strong-route.directive.spec.ts]: nit: never liked testing truthyness. Why not test it is an isntance of strong route directive class? Gives more informative error if/when test fails.

File: src/app/interfaces/strongRoute.ts [hunk 24]
```diff
@@ -1,152 +1,311 @@
...
+   * List of StrongRoutes which have isParameter set to true in the hierarchy
+   * from the root StrongRoute, to the current StrongRoute
+   * (ie. `[":siteId", ":projectId"]`)
+   */
   private readonly parameters: StrongRoute[];
+  /**
+   * List of all StrongRoutes in the hierarchy going from the root StrongRoute, to
+   * the current StrongRoute
+   */
   private readonly full: StrongRoute[];
```
Comment (atruskie) [inline src/app/interfaces/strongRoute.ts]: these two fields seem redundant. We're implementing a linked list  here, why do we need to more than double memory usage by storing references to all items in the list in another array

**Review by atruskie — APPROVED** [review]
